proposals/1141_audit-logging/README.md

```diff
@@ -0,0 +1,323 @@
+
+### Specification / How it Solves the Use Cases
+
+This proposal recommends following the recommendation of the Kubernetes docs to create a separate Pod container log.
+This new container log will contain the audit logs (and only the audit logs).
+
+#### API Changes
+
+##### Configuration Options
+
+There will be very few user-facing configuration options for audit logging in the first version of the feature. If later
+found to be needed, more configuration could be added in future versions.
+
+This proposal recommends adding a single on/off install-time configuration option for disabling audit logs. By default,
+audit logs will be enabled. An admin user who is concerned about logging identities, for example because usernames may
+be considered PII, may disable audit logging.
+
+Like other install-time configuration options, this option would appear in the values.yaml file of the Supervisor and
+Concierge deployment directories. The selected value would be rendered into the "static" ConfigMap, and read by the
+Supervisor or Concierge app's Golang code at Pod startup time.
+
+##### Event Data
+
+Deciding every specific audit event is an implementation detail beyond the scope of this proposal.
+
+Generally, the following data should be included with every audit event, whenever possible:
+
+- What type of event occurred (e.g. login)
+- Outcomes of event (succeed or fail)
+- When the event occurred
+- Where the event occurred (Kubernetes Pod logs automatically include the ID of the Pod, which should be sufficient)
+- Source of the event (e.g. requester IP address)
+- The identity of individuals or subjects associated with the event (who initiated, who participated. etc.)
+- Details involving any objects accessed
+
+The Supervisor's audit logs would include events such as:
+
+- Upstream logins for all IdP types (started, succeeded, failed)
+- Upstream refresh for all IdP types (succeeded, failed)
+- Upstream group refresh for all IdP types (succeeded, failed)
+- Downstream login (started, succeeded, failed)
+- Downstream token exchange (succeeded, failed)
+- Session expired
+- Maybe: The equivalent of access log events for all Supervisor endpoints, since there is no other component providing
+  access logs. This would include logging things like calls to the Supervisor's OIDC well-known discovery endpoint.
+  These logs could help an investigator determine more about the usage pattern of a suspicious client.
+- Maybe: Newly authenticated user is associated with “admin” RBAC. Note that the Supervisor is not directly aware of
+  RBAC, so determining this would require otherwise unnecessary calls to the Kubernetes API server, which would degrade
+  the performance of the Supervisor. It's also not clear what would constitute "admin" level access, since RBAC is
+  configurable at a very fine-grained level. On the other hand, the Supervisor is directly aware of the user's group
+  memberships, which could be logged.
```

Comment:
This is less meaningful of a check for the supervisor to perform since it is the workload clusters that are generally interacted with - i.e. this is more meaningful in the concierge, but I think it is probably desirable to do it in both places since being admin on the supervisor cluster is extra scary.

"cluster admin" is easy to define, it is a user that is allowed to perform `verbs=*` `groups=*` `resources=*` via a subject access review.  I think that is good enough for our use case.

One REST call on login will not have any meaningful impact on the supervsior, and we need to cite actual data before making decisions based on performance.  This also does not need to be performed inline with the request.  We perform inline SARs with basically every request made to the impersonation proxy and have not heard any complaints related to performance there.

Group memberships should always be audit logged, but that does not give the auditor an easy way to understand when an admin is using a cluster.

I do not see this as a maybe since it would be easy to implement and provide valuable insight to an auditor.

Reply:
> This is less meaningful of a check for the supervisor to perform since it is the workload clusters that are generally interacted with

That's a good point. The RBAC permissions on the Supervisor's cluster will almost certainly be different than the workload clusters where the user is actually logging in. Additionally, if the Concierge isn't running on the same cluster as the Supervisor, then the user may not even be able to use the Supervisor to log in to the Supervisor's own cluster. It seems hard to say that such a check would have any particular meaning.

To me, this check would make a lot more sense on the Concierge, which is where the actual "login" to the cluster is happening. In the Concierge, the RBAC permissions are known for that specific cluster and could be checked. If that Concierge happens to be running on the same cluster as the Supervisor, then it takes care of that use case too.

> "cluster admin" is easy to define, it is a user that is allowed to perform verbs=* groups=* resources=* via a subject access review.

What if the user doesn't have unlimited permissions but does have permission to edit resources in the Supervisor's or Concierge's namespace? That type of lesser admin might be able to change the Pinniped configuration or session data to enable privilege escalation for themselves. There's a lot of grey area in what it means to be an "admin".

However, defining it narrowly as only users with `verbs=* groups=* resources=*` permissions would at least share some useful information in the logs about those users who have unlimited admin access.

> One REST call on login will not have any meaningful impact on the supervsior

Fair point, as long as we could achieve the desired result with a small number of calls. It seems like the [API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#subjectaccessreviewspec-v1-authorization-k8s-io) only lets you check one API group, one resource, and one verb at a time (with the exception of using `*` wildcards), so if we tried to invent a more nuanced version then it could require lots of calls. The simple check that you proposed probably wouldn't have noticeable performance impact, as you said.

Reply:
> It seems hard to say that such a check would have any particular meaning in the Supervisor

It would mean "if this identity was to asserted on the cluster that the supervisor is running on, it would be an admin level identity."  This is basically always true in the standard pinniped deployment, so including it in the supervisor audit log seems prudent because it asserts that the identity is a super-cluster-admin - i.e. they are cluster-admin on all the workload clusters as well because they can influence the supervisor to do anything they want.

> What if the user doesn't have unlimited permissions but does have permission to edit resources ...

The `*` based approach is good enough for an MVP and provides value.

> if we tried to invent a more nuanced version then it could require lots of calls

The subject rules review API gives you the whole list in one API call.  I am just trying to avoid the complexity in the MVP.  I could see an API where an admin defines resources/verbs/etc that they deem to be "risky" so that we could audit based on that - again, not something I want to worry about during the MVP.

Reply:
> It would mean "if this identity was to asserted on the cluster that the supervisor is running on

The identity cannot be asserted on the Supervisor's cluster without going through TokenCredentialRequest of the Concierge which is running on the same cluster, right? So if the Concierge were the component which performed the subject access review and audit logged the results, then doesn't that cover the use case? It would also cover the workload cluster use case, when the Concierge happens to be running on a workload cluster.